geps/gep-2162.md

```diff
@@ -0,0 +1,114 @@
+# GEP-2162: Supported features in GatewayClass Status
+
+* Issue: [#2162](https://github.com/kubernetes-sigs/gateway-api/issues/2162)
+* Status: Provisional
+
+## TLDR
+
+This GEP proposes to enhance the [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185) to include a list of gw-api features supported by the installed Gateway Class. 
+
+## Goals
+
+* Enable users easily see what features the implementation(gateway class) support
+* Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+## Non-Goals
+
+* Validate correctness of supported features published by the implementation.
+
+## Introduction
+
+The current  [GatewayClassStatus](https://github.com/kubernetes-sigs/gateway-api/blob/f2cd9bb92b4ff392416c40d6148ff7f76b30e649/apis/v1beta1/gatewayclass_types.go#L185 is only used to store conditions the controller publishes.
+
+With the evolution of the Conformance tests suite and features, we want to;
+
+1. Enable users easily see what features the implementation support
+1. Improving UX and potentially provide clear errors when a user is trying to use a feature the installed GWC doesn't support
+
+This doc proposes to enhance the GatewayClassStatus API so implementations could publish a list of features they support/don't support.
+
+Implementations **should** publish the supported/unsupported features before Accepting the GatewayClass.
+
+They can either maintain static lists of supported/unsupported features OR generate if from their conformance tests results/reports as part of their CI.
```

Comment:
```suggestion
They can either maintain static lists of supported/unsupported features OR generate them from their conformance tests results/reports as part of their CI.
```

Is there any standard or guideline for the name of supported/unsupported features?

Reply:
BTW, what is the granularity of the features?
For instance, if an implementation supports TLSRoute, does it mean the implementation needs to support both TLS Termination mode & TLS passthrough mode? Or each mode should have its own feature name?

Reply:
> Is there any standard or guideline for the name of supported/unsupported features?
Constant values from [features.go](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go). for example - https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go#L118-L142

Reply:
> BTW, what is the granularity of the features?
For instance, if an implementation supports TLSRoute, does it mean the implementation needs to support both TLS Termination mode & TLS passthrough mode? Or each mode should have its own feature name?

The [documentation](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.GatewayTLSConfig) and the [code](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go#L166) don't differentiate between support for TLS Termination and passthrough. They are both under TLSCore feature.

So to answer your question, for this specific q - yes, since the feature name would be TLSRoute is one feature.

A different example would be like HTTPRedirect where we have[ different feature names](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/utils/suite/features.go#L126-L133) for different sub capabilities